File: apps/web/core/api/task-gateway.ts
import type {
  AssignTaskRequest,
  AttachTaskFileRequest,
  CompleteTaskRequest,
  CreateManualTaskRequest,
  DismissTaskRequest,
  EditTaskRequest,
  RescheduleTaskRequest,
  Task,
  TaskActivityListResult,
  TaskAttachment,
  TaskListResult,
  TaskStatus,
} from '@verdery/api-contracts';
import { IDEMPOTENCY_KEY_HEADER, IF_MATCH_HEADER } from '@verdery/api-contracts';

import type { ApiClient } from './client';
import { csrfHeader } from './csrf';
import type { ApiResult } from './result';

export interface TaskGateway {
  create(
    gardenId: string,
    input: CreateManualTaskRequest,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  list(
    gardenId: string,
    statuses: readonly TaskStatus[] | null,
    signal?: AbortSignal,
  ): Promise<ApiResult<TaskListResult>>;
  edit(
    gardenId: string,
    taskId: string,
    input: EditTaskRequest,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  reschedule(
    gardenId: string,
    taskId: string,
    input: RescheduleTaskRequest,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  complete(
    gardenId: string,
    taskId: string,
    input: CompleteTaskRequest,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  dismiss(
    gardenId: string,
    taskId: string,
    input: DismissTaskRequest,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  skip(
    gardenId: string,
    taskId: string,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  delete(
    gardenId: string,
    taskId: string,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  attachFile(
    gardenId: string,
    taskId: string,
    input: AttachTaskFileRequest,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<TaskAttachment>>;
  assign(
    gardenId: string,
    taskId: string,
    input: AssignTaskRequest,
    expectedRevision: number,
    idempotencyKey: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<Task>>;
  activity(
    gardenId: string,
    taskId: string,
    signal?: AbortSignal,
  ): Promise<ApiResult<TaskActivityListResult>>;
}

function revisionHeaders(expectedRevision: number, idempotencyKey: string): Record<string, string> {
  return {
    [IDEMPOTENCY_KEY_HEADER]: idempotencyKey,
    [IF_MATCH_HEADER]: `"${String(expectedRevision)}"`,
    ...csrfHeader(),
  };
}

function listQuery(statuses: readonly TaskStatus[] | null): string {
  if (statuses === null || statuses.length === 0) {
    return '';
  }
  return `?status=${statuses.map(encodeURIComponent).join(',')}`;
}

/**
 * Gateway for the tasks-recommendations manual-task endpoints.
 *
 * `attachFile` is implemented for contract completeness and covered by
 * `task-gateway.test.ts`, but no `features/tasks` hook or component calls it
 * this pass — the same media-upload gap `plant-gateway.ts`'s module doc
 * comment explains. `deleteTask` is a status transition (`POST`, not HTTP
 * `DELETE`) per the contract's own `deleteTask` operation description; this
 * gateway's `delete` method name reflects the UI action, not the verb.
 *
 * `assign`/`activity` are P9A-TASK-01's assignment and shared-activity-history
 * operations: `assign` shares `revisionHeaders` with every other
 * revision-guarded mutation on this tag, so a stale assignment surfaces the
 * identical `412`/`error.staleRevision` conflict `edit`/`reschedule` already
 * do. `activity` carries neither header — it is a plain read, open to every
 * garden role per the operation's own contract description.
 *
 * Source: packages/api-contracts/openapi.yaml, tag `Tasks`;
 * architecture/web-application-design.md, section "8. API Access".
 */
export function createTaskGateway(client: ApiClient): TaskGateway {
  return {
    create(gardenId, input, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks`,
        body: input,
        headers: { [IDEMPOTENCY_KEY_HEADER]: idempotencyKey, ...csrfHeader() },
        ...(signal === undefined ? {} : { signal }),
      });
    },

    list(gardenId, statuses, signal) {
      return client.request<TaskListResult>({
        method: 'GET',
        path: `/gardens/${gardenId}/tasks${listQuery(statuses)}`,
        ...(signal === undefined ? {} : { signal }),
      });
    },

    edit(gardenId, taskId, input, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'PATCH',
        path: `/gardens/${gardenId}/tasks/${taskId}`,
        body: input,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    reschedule(gardenId, taskId, input, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/reschedule`,
        body: input,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    complete(gardenId, taskId, input, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/complete`,
        body: input,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    dismiss(gardenId, taskId, input, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/dismiss`,
        body: input,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    skip(gardenId, taskId, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/skip`,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    delete(gardenId, taskId, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/delete`,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    attachFile(gardenId, taskId, input, idempotencyKey, signal) {
      return client.request<TaskAttachment>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/attachments`,
        body: input,
        headers: { [IDEMPOTENCY_KEY_HEADER]: idempotencyKey, ...csrfHeader() },
        ...(signal === undefined ? {} : { signal }),
      });
    },

    assign(gardenId, taskId, input, expectedRevision, idempotencyKey, signal) {
      return client.request<Task>({
        method: 'POST',
        path: `/gardens/${gardenId}/tasks/${taskId}/assign`,
        body: input,
        headers: revisionHeaders(expectedRevision, idempotencyKey),
        ...(signal === undefined ? {} : { signal }),
      });
    },

    activity(gardenId, taskId, signal) {
      return client.request<TaskActivityListResult>({
        method: 'GET',
        path: `/gardens/${gardenId}/tasks/${taskId}/activity`,
        ...(signal === undefined ? {} : { signal }),
      });
    },
  };
}

File: apps/ios/Sources/CoreNetworking/TaskGateway.swift
import CoreDomain
import CoreObservability
import Foundation

/// The application's view of the manual task operations.
///
/// Features depend on this protocol, never on `URLSession` or a generated
/// client, so a feature test needs no network and no server — the same
/// reason `GardenGateway` exists.
///
/// Source: architecture/ios-application-design.md, section "9. Networking";
/// packages/api-contracts/openapi.yaml, tag `Tasks`.
public protocol TaskGateway: Sendable {
    func createManualTask(
        gardenId: String,
        targetKind: TaskTargetKind,
        targetGardenAreaMapObjectId: String?,
        targetPlantId: String?,
        title: String,
        notes: String?,
        dueDate: String?,
        timeWindowStart: Date?,
        timeWindowEnd: Date?,
        urgency: TaskUrgency?,
        originObservationId: String?,
        idempotencyKey: String
    ) async throws -> GardenTask

    /// `statuses` empty returns every status — the contract's own default
    /// for an omitted filter.
    func listTasksForGarden(gardenId: String, statuses: [TaskStatus]) async throws -> [GardenTask]

    func editTask(
        gardenId: String,
        taskId: String,
        title: String?,
        notes: FieldUpdate<String>,
        dueDate: FieldUpdate<String>,
        timeWindowStart: FieldUpdate<Date>,
        timeWindowEnd: FieldUpdate<Date>,
        urgency: TaskUrgency?,
        recurrenceRule: FieldUpdate<String>,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask

    func rescheduleTask(
        gardenId: String,
        taskId: String,
        dueDate: FieldUpdate<String>,
        timeWindowStart: FieldUpdate<Date>,
        timeWindowEnd: FieldUpdate<Date>,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask

    func completeTask(
        gardenId: String,
        taskId: String,
        completionNote: String?,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask

    func dismissTask(
        gardenId: String,
        taskId: String,
        reason: String?,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask

    func skipTask(gardenId: String, taskId: String, expectedRevision: Int, idempotencyKey: String) async throws -> GardenTask

    /// A status transition (`'deleted'`), not a hard delete — see
    /// `packages/api-contracts/openapi.yaml`'s own description of
    /// `deleteTask`. Named to match the contract's operation, not to imply
    /// the row disappears.
    func deleteTask(gardenId: String, taskId: String, expectedRevision: Int, idempotencyKey: String) async throws -> GardenTask

    /// See `FeatureTasks`'s doc comment on the (deliberately absent)
    /// attachment UI for why this method has no entry point this pass, even
    /// though it is fully implemented and tested here.
    func attachTaskFile(
        gardenId: String,
        taskId: String,
        mediaId: String,
        idempotencyKey: String
    ) async throws -> TaskAttachment
}

/// URLSession-backed implementation of the manual task operations.
public struct URLSessionTaskGateway: TaskGateway {
    private let transport: HTTPTransport

    public init(
        configuration: APIConfiguration,
        session: URLSession = .shared,
        correlationIdentifiers: any CorrelationIdentifierProvider =
            RandomCorrelationIdentifierProvider(),
        authTokenProvider: any AuthTokenProvider,
        appCheckTokenProvider: (any AppCheckTokenProvider)? = nil,
        log: any DiagnosticLog = NoOperationDiagnosticLog()
    ) {
        self.transport = HTTPTransport(
            configuration: configuration,
            session: session,
            correlationIdentifiers: correlationIdentifiers,
            authTokenProvider: authTokenProvider,
            appCheckTokenProvider: appCheckTokenProvider,
            log: log
        )
    }

    public func createManualTask(
        gardenId: String,
        targetKind: TaskTargetKind,
        targetGardenAreaMapObjectId: String?,
        targetPlantId: String?,
        title: String,
        notes: String?,
        dueDate: String?,
        timeWindowStart: Date?,
        timeWindowEnd: Date?,
        urgency: TaskUrgency?,
        originObservationId: String?,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks",
            body: CreateManualTaskRequestTransport(
                target: TaskTargetRequestTransport(
                    kind: targetKind,
                    gardenAreaMapObjectId: targetGardenAreaMapObjectId,
                    plantId: targetPlantId
                ),
                title: title,
                notes: notes,
                dueDate: dueDate,
                timeWindow: Self.timeWindowTransport(start: timeWindowStart, end: timeWindowEnd),
                urgency: urgency,
                originObservationId: originObservationId
            ),
            headers: [APIConfiguration.idempotencyKeyHeader: idempotencyKey],
            acceptedStatusCodes: [201]
        )
        return result.domainValue
    }

    public func listTasksForGarden(gardenId: String, statuses: [TaskStatus]) async throws -> [GardenTask] {
        var path = "gardens/\(gardenId)/tasks"
        if !statuses.isEmpty {
            path += "?status=" + statuses.map(\.rawValue).joined(separator: ",")
        }

        let result: TaskListResultTransport = try await transport.get(
            operationPath: path,
            acceptedStatusCodes: [200]
        )
        return result.items.map(\.domainValue)
    }

    public func editTask(
        gardenId: String,
        taskId: String,
        title: String?,
        notes: FieldUpdate<String>,
        dueDate: FieldUpdate<String>,
        timeWindowStart: FieldUpdate<Date>,
        timeWindowEnd: FieldUpdate<Date>,
        urgency: TaskUrgency?,
        recurrenceRule: FieldUpdate<String>,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "PATCH",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)",
            body: EditTaskRequestTransport(
                title: title,
                notes: notes,
                dueDate: dueDate,
                timeWindow: Self.timeWindowTransport(fieldUpdateStart: timeWindowStart, fieldUpdateEnd: timeWindowEnd),
                urgency: urgency,
                recurrenceRule: recurrenceRule
            ),
            headers: revisionHeaders(expectedRevision: expectedRevision, idempotencyKey: idempotencyKey),
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }

    public func rescheduleTask(
        gardenId: String,
        taskId: String,
        dueDate: FieldUpdate<String>,
        timeWindowStart: FieldUpdate<Date>,
        timeWindowEnd: FieldUpdate<Date>,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)/reschedule",
            body: RescheduleTaskRequestTransport(
                dueDate: dueDate,
                timeWindow: Self.timeWindowTransport(fieldUpdateStart: timeWindowStart, fieldUpdateEnd: timeWindowEnd)
            ),
            headers: revisionHeaders(expectedRevision: expectedRevision, idempotencyKey: idempotencyKey),
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }

    public func completeTask(
        gardenId: String,
        taskId: String,
        completionNote: String?,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)/complete",
            body: CompleteTaskRequestTransport(completionNote: completionNote),
            headers: revisionHeaders(expectedRevision: expectedRevision, idempotencyKey: idempotencyKey),
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }

    public func dismissTask(
        gardenId: String,
        taskId: String,
        reason: String?,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)/dismiss",
            body: DismissTaskRequestTransport(reason: reason),
            headers: revisionHeaders(expectedRevision: expectedRevision, idempotencyKey: idempotencyKey),
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }

    public func skipTask(
        gardenId: String,
        taskId: String,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)/skip",
            headers: revisionHeaders(expectedRevision: expectedRevision, idempotencyKey: idempotencyKey),
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }

    public func deleteTask(
        gardenId: String,
        taskId: String,
        expectedRevision: Int,
        idempotencyKey: String
    ) async throws -> GardenTask {
        let result: GardenTaskTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)/delete",
            headers: revisionHeaders(expectedRevision: expectedRevision, idempotencyKey: idempotencyKey),
            acceptedStatusCodes: [200]
        )
        return result.domainValue
    }

    public func attachTaskFile(
        gardenId: String,
        taskId: String,
        mediaId: String,
        idempotencyKey: String
    ) async throws -> TaskAttachment {
        let result: TaskAttachmentTransport = try await transport.send(
            method: "POST",
            operationPath: "gardens/\(gardenId)/tasks/\(taskId)/attachments",
            body: AttachTaskFileRequestTransport(mediaId: mediaId),
            headers: [APIConfiguration.idempotencyKeyHeader: idempotencyKey],
            acceptedStatusCodes: [201]
        )
        return result.domainValue
    }

    private func revisionHeaders(expectedRevision: Int, idempotencyKey: String) -> [String: String] {
        [
            APIConfiguration.idempotencyKeyHeader: idempotencyKey,
            APIConfiguration.ifMatchHeader: "\"\(expectedRevision)\"",
        ]
    }

    /// Builds a `timeWindow` object for a create request from plain
    /// optionals — `nil` for both means "omit `timeWindow` entirely," since a
    /// create has no existing window to distinguish "leave unchanged" from.
    private static func timeWindowTransport(start: Date?, end: Date?) -> TaskTimeWindowRequestTransport? {
        guard start != nil || end != nil else { return nil }
        return TaskTimeWindowRequestTransport(start: .set(start), end: .set(end))
    }

    /// Builds a `timeWindow` object for an edit/reschedule request from
    /// ``FieldUpdate``s — `.unchanged` for both means "omit `timeWindow`
    /// entirely," leaving the existing window untouched.
    private static func timeWindowTransport(
        fieldUpdateStart: FieldUpdate<Date>,
        fieldUpdateEnd: FieldUpdate<Date>
    ) -> TaskTimeWindowRequestTransport? {
        if case .unchanged = fieldUpdateStart, case .unchanged = fieldUpdateEnd { return nil }
        return TaskTimeWindowRequestTransport(start: fieldUpdateStart, end: fieldUpdateEnd)
    }
}
